applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -300,23 +351,206 @@ static int handle_at_slmuart(const char *at_cmd, uint32_t *baudrate)
+{
+	int ret;
+	uint16_t size_limit;
+	uint16_t time_limit;
+
+	ret = at_parser_params_from_str(at_cmd, NULL, &at_param_list);
+	if (ret) {
+		LOG_ERR("Failed to parse AT command %d", ret);
+		return -EINVAL;
+	}
+	switch (at_parser_cmd_type_get(at_cmd)) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		ret = at_params_short_get(&at_param_list, 1, &size_limit);
+		if (ret) {
+			return ret;
+		}
+		ret = at_params_short_get(&at_param_list, 2, &time_limit);
+		if (ret) {
+			return ret;
+		}
+		if (datamode_time_limit < DATAMODE_TIME_LIMIT_MAX &&
+		    datamode_size_limit < DATAMODE_SIZE_LIMIT_MAX) {
+			datamode_size_limit = size_limit;
+			datamode_time_limit = time_limit;
+		} else {
+			return -EINVAL;
+		}
+		break;
+
+	case AT_CMD_TYPE_READ_COMMAND:
+		sprintf(rsp_buf, "#XDATACTRL: %d, %d\r\n",
+				datamode_size_limit, datamode_time_limit);
+		rsp_send(rsp_buf, strlen(rsp_buf));
 		ret = 0;
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf, "#XDATACTRL=<size_limit>,<time_limit>\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		ret = 0;
+		break;
+
+	default:
+		break;
 	}
 
 	return ret;
 }
 
+static void raw_send(struct k_work *work)
+{
+	int err;
+
+	ARG_UNUSED(work);
+
+	LOG_HEXDUMP_DBG(at_buf, at_buf_len, "RX");
+
+	if (slm_tcp_get_datamode()) {
+		(void)slm_tcp_send_datamode(at_buf, at_buf_len);
+		goto sent;
+	}
+
+	if (slm_udp_get_datamode()) {
+		(void)slm_udp_send_datamode(at_buf, at_buf_len);
+		goto sent;
+	}
+
+sent:
+	err = uart_rx_enable(uart_dev, uart_rx_buf[0],
+			     sizeof(uart_rx_buf[0]), UART_RX_TIMEOUT);
+	if (err) {
+		LOG_ERR("UART RX failed: %d", err);
+	}
+	rx_start = k_uptime_get();
+	at_buf_len = 0;
+}
+
+static void inactivity_timer_handler(struct k_timer *timer)
+{
+	ARG_UNUSED(timer);
+
+	LOG_INF("time limit reached, size: %d", at_buf_len);
+
+	if (at_buf_len > 0) {
+		uart_rx_disable(uart_dev);
+		k_work_submit(&raw_send_work);
+	}
+}
+
+K_TIMER_DEFINE(inactivity_timer, inactivity_timer_handler, NULL);
+
+static void silence_timer_handler(struct k_timer *timer)
+{
+	ARG_UNUSED(timer);
+
+	/* Drop pending data unconditionally */
+	if (datamode_time_limit > 0) {
+		k_timer_stop(&inactivity_timer);
+	}
+	at_buf_len = 0;
+
+	/* quit datamode */
+	exit_datamode();
+	if (slm_tcp_get_datamode()) {
+		slm_tcp_set_datamode_off();
+	}
+	if (slm_udp_get_datamode()) {
+		slm_udp_set_datamode_off();
+	}
+	datamode_off_pending = false;
+}
+
+K_TIMER_DEFINE(silence_timer, silence_timer_handler, NULL);
+
+static int raw_rx_handler(const uint8_t *data, int datalen)
+{
+	const char *QUIT = CONFIG_SLM_DATAMODE_TERMINATOR;
+	int LEN = strlen(QUIT);
+	int64_t SILENCE = CONFIG_SLM_DATAMODE_SILENCE * MSEC_PER_SEC;
+
+	/* First, check conditions for quiting datamode
+	 */
+	if (SILENCE > k_uptime_delta(&rx_start)) {
+		if (datamode_off_pending) {
+			/* quit procedure aborted */
+			k_timer_stop(&silence_timer);
+			datamode_off_pending = false;
+			memmove(at_buf + at_buf_len, QUIT, LEN);
```

Comment:
Ok, now I'm a bit confused, previous implementation  placed the terminator at the beginning of the at_buf, but now it's placed at the end. If the latter is correct, just use `memcpy()` (`memmove()` is good to move data within the same buffer, here we just copy the data).

Reply:
Previous SLM send to LTE upon each UART RX. Now time/size limit for data mode is required so SLM has to buffer what's received. In the case of cancelling quit by "+++", we APPEND this string to the buffer then append new RX data down below.